staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -776,7 +778,8 @@ func (s *store) GetList(ctx context.Context, key string, opts storage.ListOption
 			obj, err := decodeListItem(ctx, data, uint64(kv.ModRevision), s.codec, s.versioner, newItemFunc)
 			if err != nil {
 				recordDecodeError(s.groupResourceString, string(kv.Key))
-				return err
+				failedKeys[string(kv.Key)] = err
```

Comment:
In an ideal world undecryptable data can be recovered if the key becomes accessible again whereas corrupted data is completely lost. The only way to fix any residual problem is to delete the data completely and recreate it manually. Wouldn't it make sense to handle both of these scenarios separately to highlight that difference to the users?

Reply:
> In an ideal world undecryptable data can be recovered if the key becomes accessible again whereas corrupted data is completely lost.

Scenario: KAS and etcd communicate over a network with aging hardware and a bit gets flipped every now and then. This way your data may be temporarily corrupt.
This enhancement is mostly meant to unblock execution by a deletion of undecryptable/corrupt resources that can be automatically recovered (e.g. recreated by a controller), or are no longer needed - see the [user stories in the enhancement](https://github.com/kubernetes/enhancements/blob/master/keps/sig-auth/3926-handling-undecryptable-resources/README.md#user-stories-optional).

The transformer in this part of the codebase is a generic construct and may not be exclusive to encryption - even though that's probably the only thing using it today. A more generic error seems appropriate.

We could probably special-case just the encryption error in the specific transformer and have a special API error just for that, but I can't say that I'm a big fan of such an approach.

Reply:
> The transformer in this part of the codebase is a generic construct and may not be exclusive to encryption - even though that's probably the only thing using it today. A more generic error seems appropriate.

That's a good point. I was too focused on the encryption use-case which made it seems like we were handling two completely different errors in the same way. But I think it is fine to have both of them reported the same way and then users can rely on the Status Causes to figure out what is exactly happening.